hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java

```diff
@@ -360,16 +355,108 @@ public ContainerInfo getMatchingContainer(final long sizeRequired,
       String owner, Pipeline pipeline) {
     try {
       //TODO: #CLUTIL See if lock is required here
-      return containerStateManager
-          .getMatchingContainer(sizeRequired, owner, pipelineManager,
-              pipeline);
+      NavigableSet<ContainerID> containerIDs =
+          pipelineManager.getContainersInPipeline(pipeline.getId());
+      if (containerIDs == null) {
+        LOG.error("Container list is null for pipeline=", pipeline.getId());
+        return null;
+      }
+
+      containerIDs = getContainersForOwner(containerIDs, owner);
+      if (containerIDs.size() < numContainerPerOwnerInPipeline) {
+        synchronized (pipeline) {
+          // TODO: #CLUTIL Maybe we can add selection logic inside synchronized
+          // as well
+          containerIDs = getContainersForOwner(
+              pipelineManager.getContainersInPipeline(pipeline.getId()), owner);
+          if (containerIDs.size() < numContainerPerOwnerInPipeline) {
+            ContainerInfo containerInfo =
+                containerStateManager.allocateContainer(pipelineManager, owner,
+                    pipeline);
+            // Add to DB
+            addContainerToDB(containerInfo);
+            containerStateManager.updateLastUsedMap(pipeline.getId(),
+                containerInfo.containerID(), owner);
+            return containerInfo;
+          }
+        }
+      }
+
+      ContainerInfo containerInfo =
+          containerStateManager.getMatchingContainer(sizeRequired, owner,
+              pipeline.getId(), containerIDs);
+      if (containerInfo == null) {
+        synchronized (pipeline) {
+          containerInfo =
+              containerStateManager.allocateContainer(pipelineManager, owner,
+                  pipeline);
+          // Add to DB
+          addContainerToDB(containerInfo);
+          containerStateManager.updateLastUsedMap(pipeline.getId(),
+              containerInfo.containerID(), owner);
+        }
+      } else {
+        containerStateManager.updateLastUsedMap(pipeline.getId(),
+            containerInfo.containerID(), owner);
+      }
+      // TODO: #CLUTIL cleanup entries in lastUsedMap
```

Comment:
containerStateManager.updateLastUsedMap can be called just before the return statement and avoided in multiple places.

Reply:
Done